istioctl/cmd/injector-list.go

```diff
@@ -0,0 +1,319 @@
+	for _, hook := range hooks {
+		revision := hook.ObjectMeta.GetLabels()[label.IstioRev]
+		namespaces := getMatchingNamespaces(&hook, namespaces)
+		if len(namespaces) == 0 {
+			fmt.Fprintf(w, "%s\t%s\t%s\t%s\n", "DOES NOT AUTOINJECT", hook.Name, revision, injectedImages[revision])
+			continue
+		}
+		for _, namespace := range namespaces {
+			fmt.Fprintf(w, "%s\t%s\t%s\t%s\n", namespace.Name, hook.Name, revision, injectedImages[revision])
+		}
+	}
+	return w.Flush()
+}
+
+func getInjector(namespace *v1.Namespace, hooks []admit_v1.MutatingWebhookConfiguration) *admit_v1.MutatingWebhookConfiguration {
+	// find matching hook
+	for _, hook := range hooks {
+		for _, webhook := range hook.Webhooks {
+			nsSelector, err := metav1.LabelSelectorAsSelector(webhook.NamespaceSelector)
+			if err != nil {
+				continue
+			}
+			if nsSelector.Matches(api_pkg_labels.Set(namespace.ObjectMeta.Labels)) {
+				return &hook
+			}
+		}
+	}
+	return nil
+}
+
+func getInjectedRevision(namespace *v1.Namespace, hooks []admit_v1.MutatingWebhookConfiguration) string {
+	injector := getInjector(namespace, hooks)
+	if injector != nil {
+		return injector.ObjectMeta.GetLabels()[label.IstioRev]
+	}
+	newRev := namespace.ObjectMeta.GetLabels()[label.IstioRev]
+	oldLabel, ok := namespace.ObjectMeta.GetLabels()[analyzer_util.InjectionLabelName]
+	// If there is no istio-injection=disabled and no istio.io/rev, the namespace isn't injected
+	if newRev == "" && (ok && oldLabel == "disabled" || !ok) {
+		return ""
+	}
+	if newRev != "" {
+		return fmt.Sprintf("MISSING/%s", newRev)
+	}
+	return fmt.Sprintf("MISSING/%s", analyzer_util.InjectionLabelName)
+}
+
+func getMatchingNamespaces(hook *admit_v1.MutatingWebhookConfiguration, namespaces []v1.Namespace) []v1.Namespace {
+	retval := make([]v1.Namespace, 0)
+	for _, webhook := range hook.Webhooks {
+		nsSelector, err := metav1.LabelSelectorAsSelector(webhook.NamespaceSelector)
+		if err != nil {
+			return retval
+		}
+
+		for _, namespace := range namespaces {
+			if nsSelector.Matches(api_pkg_labels.Set(namespace.Labels)) {
+				retval = append(retval, namespace)
+			}
+		}
+	}
+	return retval
+}
+
+func getPods(client kube.ExtendedClient) (map[resource.Namespace][]v1.Pod, error) {
+	retval := map[resource.Namespace][]v1.Pod{}
+	// All pods in all namespaces
+	pods, err := client.CoreV1().Pods("").List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return retval, err
+	}
+	for _, pod := range pods.Items {
+		podList, ok := retval[resource.Namespace(pod.GetNamespace())]
+		if !ok {
+			retval[resource.Namespace(pod.GetNamespace())] = []v1.Pod{pod}
+		} else {
+			retval[resource.Namespace(pod.GetNamespace())] = append(podList, pod)
+		}
+	}
+	return retval, nil
+}
+
+// getInjectedImages() returns a map of revision->dockerimage
+func getInjectedImages(client kube.ExtendedClient) (map[string]string, error) {
+	retval := map[string]string{}
+
+	// All configs in all namespaces that are Istio revisioned
```

Comment:
why are you fetching all of them rather than the specific one holding the template?

Reply:
I do a List for all ConfigMaps in `istio-system` with an `istio.io/rev` label.

For example, on my cluster I have two deployments, a 1.7.x and a 1.8.x.  The injector ConfigMaps I need are named `istio-sidecar-injector` and `istio-sidecar-injector-18x`.  I can't request by name because I don't yet know the names.  As far as I can see there is no label "istio-injector: true" or other way to narrow my query to minimize the K8s response.